File: backend/app/schemas/document.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.pagination import Page


class DocumentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID
    knowledge_base_id: uuid.UUID
    title: str
    source_type: str
    file_name: str | None
    file_path: str | None
    mime_type: str | None
    status: str
    error_message: str | None
    created_at: datetime
    updated_at: datetime


class DocumentPage(Page[DocumentResponse]):
    """A page of documents, plus the numbers on the filter controls.

    The counts ride along rather than living behind their own endpoint: they
    describe the same filtered set the page came from, and fetching them
    separately would let the two drift apart between requests.
    """

    status_counts: dict[str, int] = Field(default_factory=dict)


class IngestionScheduledResponse(BaseModel):
    document_id: uuid.UUID
    status: str = "processing"
    message: str = "Document ingestion scheduled"


class PreparationScheduledResponse(BaseModel):
    document_id: uuid.UUID
    status: str = "processing"
    message: str = "Document is being prepared for chat"

File: frontend/src/types/document.ts
import type { DocumentStatus } from "@/lib/terminology";

import type { ISODateString, Page, UUID } from "./api";

export type { DocumentStatus };

export type DocumentSourceType = "upload" | "url" | "manual";

export interface KnowledgeDocument {
  id: UUID;
  organization_id: UUID;
  knowledge_base_id: UUID;
  title: string;
  source_type: DocumentSourceType;
  file_name: string | null;
  file_path: string | null;
  mime_type: string | null;
  status: DocumentStatus;
  error_message: string | null;
  created_at: ISODateString;
  updated_at: ISODateString;
}

/**
 * A page of documents, plus the numbers shown on the filter controls.
 *
 * The counts describe the current search but ignore the selected status, so
 * every filter chip can show how many documents it would reveal.
 */
export interface DocumentPage extends Page<KnowledgeDocument> {
  status_counts: Partial<Record<DocumentStatus, number>>;
}

export interface DocumentUpload {
  file: File;
  title?: string | null;
}

/** Response from the endpoints that schedule background work on a document. */
export interface DocumentTaskAccepted {
  document_id: UUID;
  status: string;
  message: string;
}
